webapp/channels/src/actions/hooks.js

```diff
@@ -89,3 +89,31 @@ export function runMessageWillBeUpdatedHooks(newPost, oldPost) {
         return {data: post};
     };
 }
+
+export function runDesktopNotificationHooks(post, msgProps, channel, teamId, title, body, silent, soundName, url, notify) {
+    return async (dispatch, getState) => {
+        const hooks = getState().plugins.components.DesktopNotificationHooks;
+        if (!hooks || hooks.length === 0) {
+            return {args: {title, body, silent, soundName, url, notify}};
+        }
+
+        let args = {title, body, silent, soundName, url, notify};
+        for (const hook of hooks) {
+            const result = await hook.hook(post, msgProps, channel, teamId, args); // eslint-disable-line no-await-in-loop
```

Comment:
What's the reason to make the hook async if we are going to await on it this way?

Reply:
This is tricky -- I'm not sure how to explain it succinctly (it could be a dev meeting talk). Since the redux thunk is async, and we can't restrict hooks to be non-async, we have to await here so that the result is resolved to its value (and not a promise). Even if the hook is non-async, I think (but not 100% certain) that js will change it to an async because the thunk is async. (I need to look into that.) Good question.

Reply:
Looking forward to that talk :roll_eyes: